### ISSUE TITLE
Inferred type of cls doesn't match type of class

### ISSUE THREAD

**Issue by ermishechkin** [microsoft/pyright#1214]
**Describe the bug**
Inferred type of `cls` variable in magic methods doesn't match type of class. It works fine if `cls` is annotated explicitly.

**Screenshots or Code**
```python
from __future__ import annotations
from typing import Type

class Class:
    def __init_subclass__(cls, *, arg: int) -> None:
        func(cls, arg)

    def __new__(cls) -> Class:
        func(cls, 9)
        return super().__new__(cls)

def func(klass: Type[Class], arg: int):
    pass
```

```
6:14 - error: Argument of type "Type[Class]" cannot be assigned to parameter "klass" of type "Type[Class]" in function "func"
  Type "Type[Class]" cannot be assigned to type "Class" (reportGeneralTypeIssues)
9:14 - error: Argument of type "Type[Class]" cannot be assigned to parameter "klass" of type "Type[Class]" in function "func"
  Type "Type[Class]" cannot be assigned to type "Class" (reportGeneralTypeIssues)
```

**VS Code extension or command-line**
pyright 1.1.89
(worked fine in 1.1.88)

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is now addressed in pyright 1.1.90, which I just published. It will also be included in the next release of pylance.
